### PR TITLE
feat(mesh): edit collection

### PR DESCRIPTION
**Pull request by caesarWHLee**
# Feature
- Add Collection phase2
- Edit Collection 

## Add Collection phase 2
- add drag and drop for sorting collection picks (stories) - [dnd-kit sortable](https://docs.dndkit.com/presets/sortable)
- add filter for story source (pick, bookmark or pick and bookmark) - [figma](https://www.figma.com/design/wqODc69zqgbu7TZFtQAmnN/2024_READr-Mesh_Web?node-id=13082-325116&t=2oOh1jrn4mVqSkhX-4)

## Edit Collection 
- refactor file system (add route group to improve readability)
- add edit routes and redirect back to collection page if device is changed
  - Desktop
    - `/collection/${collectionId}/edit`
  - Mobile
    - `/collection/${collectionId}/edit-title`
    - `/collection/${collectionId}/edit-summary`
    - `/collection/${collectionId}/edit-stories`
- create `editCollectionContext` based on `createCollectionContext`

### GQL mutation
To support updating collection in multiple ways:
- update title and heroImage
- update summary
- update collection picks
- update all

The mutation `UpdateCollection` was designed to achieve the scenarios. Update will be ignore if default data is provided.
```
updateCollectionData= {}
createCollectionPicksData: []
updateCollectionPicksData: []
deleteCollectionPicksData: []
```

Also, to update collectionPicks is a bit complicated, the mutation needs to know which part is new created, which part is updated (sortOrder upadted) and which part is deleted. Check the gql mutation `UpdateCollection` and `app/collection/(mutate)/_utils/prepare-update-collection`'s `prepareUpdateCollectionPicks` function for further implementation.

## Follow-up
- Add search in the future
- Edit collection type (timeline)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208535200012070

### CODE REVIEW

File: packages/mesh/context/edit-collection.tsx [hunk 1]
```diff
@@ -1,210 +1,338 @@
...
+  )
+  const [bookmarkCandidates, setBookmarkCandidates] = useState<StoryCandidates>(
+    initialStoryCandidate
+  )
+
   const [collectionPickStories, setCollectionPickStories] = useState<
     CollectionPickStory[]
-  >([])
+  >(
+    initialCollection.collectionpicks?.reduce((acc, curr) => {
```
Comment (erase2004) [inline packages/mesh/context/edit-collection.tsx]: Ln. 112 用 `acc: CollectionPickStory[]` 即可取代 Ln. 117 的 `as`

File: packages/mesh/context/edit-collection.tsx
Comment (erase2004) [inline packages/mesh/context/edit-collection.tsx:1]: 這應該 `'use client'`

Comment (erase2004) [reply packages/mesh/context/edit-collection.tsx:1]: 重複出現的 `setCrossPageToast({ status: 'fail', text: '編輯集錦失敗，請重新嘗試' })` 可以統一處理

Comment (erase2004) [reply packages/mesh/context/edit-collection.tsx:1]: 重複出現的 `router.push(`/collection/${collectionId}`)` 可以統一處理

File: packages/mesh/app/collection/(mutate)/(edit)/_types/edit-collection.ts
Comment (erase2004) [inline packages/mesh/app/collection/(mutate)/(edit)/_types/edit-collection.ts:1]: 不太清楚為什麼要用 `TypeXXX` 來命名

File: packages/mesh/app/collection/(mutate)/_components/sort-story-card.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    transition,
+    isDragging,
+  } = useSortable({ id: story.id })
+
+  const { setCollectionPickStories } = useCollection()
+
+  const style = {
+    transform: CSS.Transform.toString(transform),
+    transition,
+    zIndex: isDragging ? 999 : 'auto', // 设置拖拽中的元素在最顶层
```
Comment (erase2004) [inline packages/mesh/app/collection/(mutate)/_components/sort-story-card.tsx]: 簡體中文？

File: packages/mesh/app/collection/(mutate)/_components/pick-stories.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import InfiniteCollectionPicks from './infinite-collection-picks'
+import StoryFilter from './story-filter'
+
+export default function PickStories({
+  useCollection,
+}: {
+  useCollection: UseCollection
+}) {
+  const [isLoading, setIsLoading] = useState(false)
+  const [mounted, setMounted] = useState(false)
```
Comment (erase2004) [inline packages/mesh/app/collection/(mutate)/_components/pick-stories.tsx]: 這應該可以改用 `useRef`

**Review by erase2004 — COMMENTED** [review]
* `(mobile)/edit-stories`、`(mobile)/edit-summary`、`(mobile)/edit-title` 底下的 layout 和 page 頗為相似，應該可以把 path 整合為 `(mobile)/[action]` 來處理
* 應該建立一個 base type 讓 `CreateCollectionContextValue` 和 `EditCollectionContextValue` 去擴充，base type 會作為 `UseCollection` 這個 type 中的 return value，確保 `CreateCollectionContext` 或 `EditCollectionContext` 各自的修改，不會導致使用 `UseCollection` 相關元件出現問題

File: packages/mesh/app/collection/(mutate)/_components/pick-stories.tsx (hunk 3 above)
Comment (caesarWHLee) [reply packages/mesh/app/collection/(mutate)/_components/pick-stories.tsx]: 修改在 [a244f55](https://github.com/readr-media/Sachiel/pull/1016/commits/a244f55fcf57bd0e448ba4ec045faa6c4e68b899)

File: packages/mesh/app/collection/(mutate)/_components/sort-story-card.tsx (hunk 2 above)
Comment (caesarWHLee) [reply packages/mesh/app/collection/(mutate)/_components/sort-story-card.tsx]: 移除在 [a244f55](https://github.com/readr-media/Sachiel/pull/1016/commits/a244f55fcf57bd0e448ba4ec045faa6c4e68b899)

File: packages/mesh/app/collection/(mutate)/(edit)/_types/edit-collection.ts
Comment (caesarWHLee) [reply packages/mesh/app/collection/(mutate)/(edit)/_types/edit-collection.ts:1]: 原本是為了要和 /app/collection/(mutate)/new/_types/create-collection.ts 中的 enum 名稱保持一制性，不過想想好像有點不必要，已移除在 [a244f55](https://github.com/readr-media/Sachiel/pull/1016/commits/a244f55fcf57bd0e448ba4ec045faa6c4e68b899)
```
export enum MobileCreateCollectionStep {
  Step1SelectStories = 'mobile-step-1-select-stories',
  Step2SetTitle = 'mobile-step-2-set-title',
  Step3SetSummary = 'mobile-step-3-set-summary',
  Step4SortStories = 'mobile-step-4-sort-stoires',
}

export enum DesktopCreateCollectionStep {
  Step1EditAll = 'desktop-step-1-edit-all',
  Step2SortStories = 'desktop-step-2-sort-stories',
}
```

File: packages/mesh/context/edit-collection.tsx (hunk 1 above)
Comment (caesarWHLee) [reply packages/mesh/context/edit-collection.tsx]: 修改在 [d451dff](https://github.com/readr-media/Sachiel/pull/1016/commits/d451dff733783f417f177cd4f635ecd78a567f59)

File: packages/mesh/context/edit-collection.tsx
Comment (caesarWHLee) [reply packages/mesh/context/edit-collection.tsx:1]: 修改在 [d451dff](https://github.com/readr-media/Sachiel/pull/1016/commits/d451dff733783f417f177cd4f635ecd78a567f59)

**Comment by caesarWHLee** [comment]
> * `(mobile)/edit-stories`、`(mobile)/edit-summary`、`(mobile)/edit-title` 底下的 layout 和 page 頗為相似，應該可以把 path 整合為 `(mobile)/[action]` 來處理

感謝建議，整合 edit collection 所有的 routes 在 [7bd5aaf](https://github.com/readr-media/Sachiel/pull/1016/commits/7bd5aafb3ad893c333ff51615b7e0b3ecf9c703c)

>     * 應該建立一個 base type 讓 `CreateCollectionContextValue` 和 `EditCollectionContextValue` 去擴充，base type 會作為 `UseCollection` 這個 type 中的 return value，確保 `CreateCollectionContext` 或 `EditCollectionContext` 各自的修改，不會導致使用 `UseCollection` 相關元件出現問題

新增 `BaseMutateCollectionContextValue` 讓 `CreateCollectionContextValue` 和 `EditCollectionContextValue` 去擴充 [50309d4](https://github.com/readr-media/Sachiel/pull/1016/commits/50309d48b5da746099ca026428d1a0b3f1c9a14f)

**Review by erase2004 — COMMENTED** [review]
LGTM 👏

**Review by erase2004 — APPROVED** [review]

**Pull request merged**